airflow/lineage/hook.py

```diff
@@ -106,7 +130,13 @@ def add_input_dataset(
             scheme=scheme, uri=uri, dataset_kwargs=dataset_kwargs, dataset_extra=dataset_extra
         )
         if dataset:
-            self.inputs.append((dataset, context))
+            key = self._generate_key(dataset, context)
+            if key not in self._inputs:
+                self._inputs[key] = (dataset, context)
+                self._input_counts[key] = 1
+            else:
+                self._input_counts[key] += 1
+                self._inputs[key] = (dataset, context)
```

Comment:
If we go with `defaultdict(int)` in `__init__`:

```
self._input_counts: dict[str, int] = defaultdict(int)
self._output_counts: dict[str, int] = defaultdict(int)
```
then we could probably skip the `if` here
```suggestion
            key = self._generate_key(dataset, context)
            self._inputs[key] = (dataset, context)
            self._input_counts[key] += 1
```

even if we don't switch to defaultdict, this can be probably be done outside of if: `self._inputs[key] = (dataset, context)`. 
I was thinking it should only be assigned again if not already in inputs, and if it's already there we could simply increase the count, but I'm not sure if we are certain that the LineageContext content hasn't changed, and that's why we should re-assign it, correct?

Reply:
Even if the context has changed we keep the same object in the same process - there would be no need to reassign.